File: src/solana_config/transformation/token_transfer.rs
use crate::solana_config::proto_codegen::confirmed_block::{
    InnerInstruction, TokenTransferInstruction,
};

#[cfg(feature = "STRING_TIMESTAMP")]
use crate::solana_config::proto_codegen::records_string_timestamp::TokenTransferRecord;

#[cfg(feature = "INT_TIMESTAMP")]
use crate::solana_config::proto_codegen::records_int_timestamp::TokenTransferRecord;

use super::transaction::CommonTableFields;

pub fn transform_to_token_transfer_record(
    common_table_fields: CommonTableFields,
    instruction_type: &str,
    program: String,
    token_transfer_instruction: Option<TokenTransferInstruction>,
    prev_instruction: Option<InnerInstruction>,
) -> Option<TokenTransferRecord> {
    pub const TOKEN_PROGRAM: &str = "spl-token";
    pub const MEMO_PROGRAM: &str = "spl-memo";
    pub const SYSTEM_PROGRAM: &str = "system";
    pub const TRANSFER_CHECKED: &str = "transferChecked";
    pub const TRANSFER_CHECKED_WITH_FEE: &str = "transferCheckedWithFee";
    pub const BURN_CHECKED: &str = "burnChecked";
    pub const MINT_TO_CHECKED: &str = "mintToChecked";
    pub const BURN: &str = "burn";
    pub const SPL_TRANSFER: &str = "spl-transfer";
    pub const SPL_TRANSFER_WITH_FEE: &str = "spl-transfer-with-fee";
    pub const MINT_TO: &str = "mintTo";
    pub const TRANSFER: &str = "transfer";

    match token_transfer_instruction {
        None => None,
        Some(t) => {
            let block_slot = common_table_fields.block_slot;
            let block_hash = common_table_fields.block_hash;
            let block_timestamp = common_table_fields.block_timestamp;
            let tx_signature = common_table_fields.tx_signature;
            /*
                "all incoming transfers must have an accompanying memo instruction right before the transfer instruction."
                source: https://spl.solana.com/token-2022/extensions#required-memo-on-transfer
            */
            let memo = match prev_instruction {
                Some(prev) if Some(MEMO_PROGRAM) == prev.program.as_deref() => {
                    match (prev.parsed_string.as_deref(), &prev.parsed_dict) {
                        (Some(parsed_string), _) if !parsed_string.is_empty() => {
                            Some(parsed_string.to_string())
                        }
                        (_, Some(parsed_dict))
                            if parsed_dict.info.as_ref().map_or(true, |s| !s.is_empty()) =>
                        {
                            parsed_dict.info.clone()
                        }
                        _ => None,
                    }
                }
                _ => None,
            };

            let source = t.source;
            let destination = t.destination;
            let authority = t.authority;
            let value = t.amount;
            let decimals = t.decimals;
            let mint = t.mint;
            let mint_authority = t.mint_authority;
            //let tx_signature = instruction.signature;
            let fee = t.fee_amount;
            let fee_decimals = t.fee_amount_decimals;
            let token_transfer_record = if program == TOKEN_PROGRAM {
                match instruction_type {
                    TRANSFER => TokenTransferRecord {
                        block_slot,
                        block_hash,
                        block_timestamp,
                        tx_signature,
                        source,
                        destination,
                        authority,
                        value,
                        decimals: None,
                        mint: None,
                        mint_authority,
                        transfer_type: Some(String::from(SPL_TRANSFER)),
                        fee,
                        fee_decimals,
                        memo,
                    },
                    TRANSFER_CHECKED => TokenTransferRecord {
                        block_slot,
                        block_hash,
                        block_timestamp,
                        tx_signature,
                        source,
                        destination,
                        authority,
                        value,
                        decimals,
                        mint,
                        mint_authority: None,
                        transfer_type: Some(String::from(SPL_TRANSFER)),
                        fee,
                        fee_decimals,
                        memo,
                    },
                    TRANSFER_CHECKED_WITH_FEE => TokenTransferRecord {
                        block_slot,
                        block_hash,
                        block_timestamp,
                        tx_signature,
                        source,
                        destination,
                        authority,
                        value,
                        decimals,
                        mint,
                        mint_authority: None,
                        transfer_type: Some(String::from(SPL_TRANSFER_WITH_FEE)),
                        fee,
                        fee_decimals,
                        memo,
                    },
                    BURN => TokenTransferRecord {
                        block_slot,
                        block_hash,
                        block_timestamp,
                        tx_signature,
                        source: None,
                        destination: None,
                        authority,
                        value,
                        decimals: None,
                        mint,
                        mint_authority: None,
                        transfer_type: Some(String::from(BURN)),
                        fee,
                        fee_decimals,
                        memo,
                    },
                    BURN_CHECKED => TokenTransferRecord {
                        block_slot,
                        block_hash,
                        block_timestamp,
                        tx_signature,
                        source: None,
                        destination: None,
                        authority,
                        value,
                        decimals,
                        mint,
                        mint_authority: None,
                        transfer_type: Some(String::from(BURN)),
                        fee,
                        fee_decimals,
                        memo,
                    },
                    MINT_TO => TokenTransferRecord {
                        block_slot,
                        block_hash,
                        block_timestamp,
                        tx_signature,
                        source: None,
                        destination: None,
                        authority: None,
                        value,
                        decimals: None,
                        mint,
                        mint_authority,
                        transfer_type: Some(String::from(MINT_TO)),
                        fee,
                        fee_decimals,
                        memo,
                    },
                    MINT_TO_CHECKED => TokenTransferRecord {
                        block_slot,
                        block_hash,
                        block_timestamp,
                        tx_signature,
                        source: None,
                        destination: None,
                        authority: None,
                        value,
                        decimals,
                        mint,
                        mint_authority,
                        transfer_type: Some(String::from(MINT_TO)),
                        fee,
                        fee_decimals,
                        memo,
                    },
                    _ => return None,
                }
            } else if program == SYSTEM_PROGRAM && instruction_type == TRANSFER {
                TokenTransferRecord {
                    block_slot,
                    block_hash,
                    block_timestamp,
                    tx_signature,
                    source,
                    destination,
                    authority,
                    decimals: None,
                    fee: None,
                    fee_decimals: None,
                    value,
                    mint: None,
                    mint_authority: None,
                    transfer_type: Some(String::from(TRANSFER)),
                    memo,
                }
            } else {
                return None;
            };

            Some(token_transfer_record)
        }
    }
}

File: storage-write/deprecated/transformation/token_transfer.go
package transformation

import (
	"strconv"
	"subscriber_rabbitmq/pbcodegen"
)

type TokenTransferData struct {
	source        *string
	destination   *string
	authority     *string
	value         *string
	fee           *uint64
	feeDecimals   *uint64
	memo          *string
	decimals      *uint64
	mint          *string
	mintAuthority *string
	transferType  *string
	txSignature   string
}

func extractTokenTransfer(program string, instruction_type string, tx_signature string, token_instruction *pbcodegen.Parsed, prev_instruction *pbcodegen.InnerInstruction) *TokenTransferData {
	/*
		NOTE: the program ID for token2022 is `TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb`
		but the program is still spl-token, like the existing spl-token (pre-2022)
	*/

	const TOKEN_PROGRAM = "spl-token" // this is the program for both types of token (spl and spl-2022)
	const MEMO_PROGRAM = "spl-memo"   // this is the program for both types of memo (v1 and v2)
	const SYSTEM_PROGRAM = "system"
	const TRANSFER_CHECKED = "transferChecked"
	const TRANSFER_CHECKED_WITH_FEE = "transferCheckedWithFee"
	const BURN_CHECKED = "burnChecked"
	const MINT_TO_CHECKED = "mintToChecked"
	//const MEMO_V1_ACCOUNT = "Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo"
	//const MEMO_V2_ACCOUNT = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr"
	BURN := "burn"
	SPL_TRANSFER := "spl-transfer"
	SPL_TRANSFER_WITH_FEE := "spl-transfer-with-fee"
	MINT_TO := "mintTo"
	TRANSFER := "transfer"

	token_instruction_generated := token_instruction.GetTokenTransferInstruction()
	if token_instruction_generated == nil {
		return nil
	}

	/*
		"all incoming transfers must have an accompanying memo instruction right before the transfer instruction."
		source: https://spl.solana.com/token-2022/extensions#required-memo-on-transfer
	*/
	var memo *string

	if prev_instruction != nil && prev_instruction.Program != nil {
		if *prev_instruction.Program == MEMO_PROGRAM {
			if prev_instruction.Data != nil && *prev_instruction.Data != "" {
				panic("Memo data should be parsed, because of jsonParsed block encoding")
			} else if prev_instruction.ParsedString != nil && *prev_instruction.ParsedString != "" {
				memo = prev_instruction.ParsedString
			} else if prev_instruction.ParsedDict != nil {
				memo = prev_instruction.ParsedDict.Info
			}
		}
	}

	var valueConv *string
	if token_instruction_generated.Amount == nil {
		valueConv = nil
	} else {
		_valueConv := strconv.FormatUint(*token_instruction_generated.Amount, 10)
		valueConv = &_valueConv
	}

	if program == TOKEN_PROGRAM {
		switch instruction_type {
		case TRANSFER:
			return &TokenTransferData{
				source:        token_instruction_generated.Source,
				destination:   token_instruction_generated.Destination,
				authority:     token_instruction_generated.Authority,
				value:         valueConv,
				decimals:      nil,
				fee:           token_instruction_generated.FeeAmount,
				feeDecimals:   token_instruction_generated.FeeAmountDecimals,
				memo:          memo,
				mint:          nil,
				mintAuthority: token_instruction_generated.MintAuthority,
				transferType:  &SPL_TRANSFER,
				txSignature:   tx_signature,
			}

		case TRANSFER_CHECKED:
			return &TokenTransferData{
				source:        token_instruction_generated.Source,
				destination:   token_instruction_generated.Destination,
				authority:     token_instruction_generated.Authority,
				value:         token_instruction_generated.TokenAmount,
				decimals:      token_instruction_generated.TokenAmountDecimals,
				fee:           token_instruction_generated.FeeAmount,
				feeDecimals:   token_instruction_generated.FeeAmountDecimals,
				memo:          memo,
				mint:          token_instruction_generated.Mint,
				mintAuthority: nil,
				transferType:  &SPL_TRANSFER,
				txSignature:   tx_signature,
			}
		case TRANSFER_CHECKED_WITH_FEE:
			return &TokenTransferData{
				source:        token_instruction_generated.Source,
				destination:   token_instruction_generated.Destination,
				authority:     token_instruction_generated.Authority,
				value:         token_instruction_generated.TokenAmount,
				decimals:      token_instruction_generated.TokenAmountDecimals,
				fee:           token_instruction_generated.FeeAmount,
				feeDecimals:   token_instruction_generated.FeeAmountDecimals,
				memo:          memo,
				mint:          token_instruction_generated.Mint,
				mintAuthority: nil,
				transferType:  &SPL_TRANSFER_WITH_FEE,
				txSignature:   tx_signature,
			}
		case BURN:
			return &TokenTransferData{
				source:        nil,
				destination:   nil,
				authority:     token_instruction_generated.Authority,
				value:         valueConv,
				decimals:      nil,
				fee:           token_instruction_generated.FeeAmount,
				feeDecimals:   token_instruction_generated.FeeAmountDecimals,
				memo:          memo,
				mint:          token_instruction_generated.Mint,
				mintAuthority: nil,
				transferType:  &BURN,
				txSignature:   tx_signature,
			}
		case BURN_CHECKED:
			return &TokenTransferData{
				source:        nil,
				destination:   nil,
				authority:     token_instruction_generated.Authority,
				value:         valueConv,
				decimals:      token_instruction_generated.Decimals,
				fee:           token_instruction_generated.FeeAmount,
				feeDecimals:   token_instruction_generated.FeeAmountDecimals,
				memo:          memo,
				mint:          token_instruction_generated.Mint,
				mintAuthority: nil,
				transferType:  &BURN,
				txSignature:   tx_signature,
			}
		case MINT_TO:
			return &TokenTransferData{
				source:        nil,
				destination:   nil,
				authority:     nil,
				value:         valueConv,
				decimals:      nil,
				fee:           token_instruction_generated.FeeAmount,
				feeDecimals:   token_instruction_generated.FeeAmountDecimals,
				memo:          memo,
				mint:          token_instruction_generated.Mint,
				mintAuthority: token_instruction_generated.MintAuthority,
				transferType:  &MINT_TO,
				txSignature:   tx_signature,
			}
		case MINT_TO_CHECKED:
			return &TokenTransferData{
				source:        nil,
				destination:   nil,
				authority:     nil,
				value:         valueConv,
				decimals:      token_instruction_generated.Decimals, // solana-etl only does this for mint-checked.
				fee:           token_instruction_generated.FeeAmount,
				feeDecimals:   token_instruction_generated.FeeAmountDecimals,
				memo:          memo,
				mint:          token_instruction_generated.Mint,
				mintAuthority: token_instruction_generated.MintAuthority,
				transferType:  &MINT_TO,
				txSignature:   tx_signature,
			}
		default:
			return nil
		}
	} else if program == SYSTEM_PROGRAM && instruction_type == TRANSFER {

		return &TokenTransferData{
			source:        token_instruction_generated.Source,
			destination:   token_instruction_generated.Destination,
			authority:     token_instruction_generated.Authority,
			value:         valueConv,
			decimals:      nil,
			fee:           nil,
			feeDecimals:   nil,
			memo:          memo,
			mint:          nil,
			mintAuthority: nil,
			transferType:  &TRANSFER,
			txSignature:   tx_signature,
		}
	} else {
		return nil
	}
}
